### DISCUSSION TITLE
getting issue in the shared credentials path

### DISCUSSION THREAD

**Discussion by GadiVikram111** [aws/aws-sdk-ruby#2877] [General]
Getting below error while creating AWS client

shared_credentials.rb:36:in `initialize': undefined method `[]' for nil:NilClass (NoMethodError)
User->aws->credentials
![image](https://github.com/aws/aws-sdk-ruby/assets/135595476/454ee51c-db85-46be-ac7f-7d708d8cebbd)

**Code**:
def uploadS3
        credentials = Aws::SharedCredentials.new(ENV['default'])
        s3 = Aws::S3::Client.new(
            region: "us-east-1",
            credentials: credentials
          )
        #Upload
        puts "sending file to s3"
        env=File.read('base/env.json')
        env = JSON.parse(env)
        time1= Time.new
        time2=time1.strftime("%Y-%m-%d-%H-%M-%S")
        filename="#{time2}-API-Automation-#{env['env_name']}-#{env['build_number']}"
        puts filename
        
        File.open('./reports/testReportst.xml', 'rb') do |file|
          puts "start uploading #{'./reports/testReportst.xml'} to s3"
          resp = s3.put_object(bucket: @@BUCKET_NAME, key: filename, body: file)
          puts resp
        end
        puts "File should be available at https://#{@BUCKET_NAME}.s3.amazonaws.com/#{filename}"
      end
end

**ERROR**:
C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/shared_credentials.rb:36:in `initialize': undefined method `[]' for nil:NilClass (NoMethodError)
        from ./send_xmltos3.rb:12:in `new'
        from ./send_xmltos3.rb:12:in `uploadS3'
        from ./send_xmltos3.rb:34:in `<main>'

**Comment by mullermp** [comment]
You shouldn't need to create an instance of shared credentials. The default profile should be loaded when you create your s3 client.

**Reply by GadiVikram111** [reply]
removed instance for shared credentials
![image](https://github.com/aws/aws-sdk-ruby/assets/135595476/a797c810-eb52-4e2e-a6d7-0cf22c6050dc)

**Getting below error**
C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sigv4-1.6.0/lib/aws-sigv4/signer.rb:701:in `fetch_credentials': undefined method `credentials' for Aws::SharedCredentials:Class (NoMethodError)
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sigv4-1.6.0/lib/aws-sigv4/signer.rb:238:in `sign_request'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/plugins/sign.rb:128:in `sign'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/plugins/sign.rb:47:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/plugins/transfer_encoding.rb:26:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/plugins/helpful_socket_errors.rb:12:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-s3-1.127.0/lib/aws-sdk-s3/plugins/s3_signer.rb:48:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-s3-1.127.0/lib/aws-sdk-s3/plugins/redirects.rb:20:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/plugins/retry_errors.rb:360:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/plugins/user_agent.rb:37:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-s3-1.127.0/lib/aws-sdk-s3/plugins/md5s.rb:31:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/plugins/http_checksum.rb:19:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/plugins/endpoint_pattern.rb:30:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/plugins/checksum_algorithm.rb:136:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-s3-1.127.0/lib/aws-sdk-s3/plugins/expect_100_continue.rb:23:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-s3-1.127.0/lib/aws-sdk-s3/plugins/bucket_name_restrictions.rb:21:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/rest/handler.rb:10:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/plugins/recursion_detection.rb:18:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-s3-1.127.0/lib/aws-sdk-s3/plugins/endpoints.rb:41:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/plugins/endpoint_discovery.rb:84:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/seahorse/client/plugins/endpoint.rb:47:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/plugins/param_validator.rb:26:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/seahorse/client/plugins/raise_response_errors.rb:16:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-s3-1.127.0/lib/aws-sdk-s3/plugins/sse_cpk.rb:24:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-s3-1.127.0/lib/aws-sdk-s3/plugins/dualstack.rb:21:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-s3-1.127.0/lib/aws-sdk-s3/plugins/accelerate.rb:43:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/plugins/checksum_algorithm.rb:111:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/plugins/jsonvalue_converter.rb:16:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/plugins/idempotency_token.rb:19:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/plugins/param_converter.rb:26:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/seahorse/client/plugins/request_callback.rb:71:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/aws-sdk-core/plugins/response_paging.rb:12:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/seahorse/client/plugins/response_target.rb:24:in `call'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.176.1/lib/seahorse/client/request.rb:72:in `send_request'
        from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/aws-sdk-s3-1.127.0/lib/aws-sdk-s3/client.rb:12684:in `put_object'
        from ./send_xmltos3.rb:28:in `block in uploadS3'
        from ./send_xmltos3.rb:26:in `open'
        from ./send_xmltos3.rb:26:in `uploadS3'
        from ./send_xmltos3.rb:34:in `<main>'

**Reply by mullermp** [reply]
I was trying to say what @alextwoods says below. If your credentials are in ~/.aws/credentials under a default profile, they will be automatically loaded. See here: https://docs.aws.amazon.com/sdk-for-ruby/v3/developer-guide/setup-config.html

**Comment by alextwoods** [comment]
The credential provider chain will automatically resolve the profile credentials - you can create your client with:

```ruby
s3 = Aws::S3::Client.new(
region: "us-east-1" # don't provide the `credentials` argument here, this will then use the credential provider chain and resolve credentials from the default profile.
)
```
